### PR TITLE
fix: recursive error on df last_result property (#468)

### LINKED ISSUE #468: RecursionError: maximum recursion depth exceeded on calling df.last_result
### 🐛 Describe the bug

 code:
 ```
 df = SmartDataframe(df=pd.DataFrame(dataframe), config={"llm": llm})
response = df.chat("Calculate the sum of the gdp of north american countries")
 ```
  
on accessing the ```df.last_result```  I'm getting RecursionError which I believe due to the recursively calls to the same setter. 

Full Trace:
 ```
df.last_result
  File "/Users/sanchit/pandas-ai/testenv/lib/python3.9/site-packages/pandasai/smart_dataframe/__init__.py", line 222, in last_result
    return self._dl.last_result
  File "/Users/sanchit/pandas-ai/testenv/lib/python3.9/site-packages/pandasai/smart_datalake/__init__.py", line 565, in last_result
    return self.last_result
  File "/Users/sanchit/pandas-ai/testenv/lib/python3.9/site-packages/pandasai/smart_datalake/__init__.py", line 565, in last_result
    return self.last_result
  File "/Users/sanchit/pandas-ai/testenv/lib/python3.9/site-packages/pandasai/smart_datalake/__init__.py", line 565, in last_result
    return self.last_result
  [Previous line repeated 2981 more times]
RecursionError: maximum recursion depth exceeded

 ```

I will open a PR shortly.

### PR DESCRIPTION
**Pull request by sandiemann**
- [x] closes #468
- [x] All [code checks passed](https://github.com/gventuri/pandas-ai/blob/main/CONTRIBUTING.md#-testing).

### Summary by CodeRabbit

**Release Notes**

- **Bug Fix**: Corrected the condition for setting `last_result` in the `chat` method of the `SmartDatalake` class.
- **Refactor**: Modified the `last_result` property and its setter in the `SmartDatalake` class to correctly handle `_last_result`.
- **Test**: Added a new test method `test_last_result_is_saved` in the `SmartDatalakeTest` class to verify the functionality of saving the last result.

> 🎉🐇
> 
> Hopping through the code, with a bug in sight,
> A rabbit coder's day, turning into night.
> With a fix and refactor, the code now shines bright,
> Tests ensure it's perfect, oh what a delight! 🌙✨

### CODE REVIEW

File: pandasai/smart_datalake/__init__.py [hunk 1]
```diff
@@ -92,6 +92,8 @@ def __init__(
         if self._config.enable_cache:
             self._cache = Cache()
 
+        self._last_result = []
```
[the remark this answers is not in the document]
Comment (sandiemann) [reply pandasai/smart_datalake/__init__.py:95]: it was pre defined as list (`_last_result: list`) & i did not change the type.
Although, I believe we only want to track the last response so string would suffice but nevertheless @gventuri can confirm.

[the remark this answers is not in the document]
Comment (gventuri) [reply pandasai/smart_datalake/__init__.py:95]: @mspronesti @sandiemann good catch, it's a leftover. In the previous refactoring I tried to allow multiple results in the same query, but turned out to be too involved, so reverted and apparently forgot that list in there!

[the remark this answers is not in the document]
Comment (sandiemann) [reply pandasai/smart_datalake/__init__.py:95]: Alright! I will change the type & initialize as string.

**Comment by sandiemann** [comment]
change to str type:
```
df.last_result
Out[4]: {'type': 'number', 'value': 20901884461056}
```

**Pull request merged**